File: benchmarks/runtime/programs/go/file-io.go
// file-io — the equivalent-semantics Go peer for the tuonelang file-io
// workload (ADR-0013's OS effect boundary). The identical sequence: per
// round, open/create/truncate a scratch file, write 15 sixteen-byte chunks
// (240 bytes), close it, reopen it for read, read it back **one byte per
// Read call** (mirroring the tuonelang program's byte-at-a-time read_byte
// crossings and the C peer's one-byte read(2) loop), close it, and remove
// it. The observable result is one round's byte count (reassigned, not
// accumulated): 240, the exit byte.
package main

import "os"

func roundTrip(chunks int64) int64 {
	path := "file-io-bench.tmp"
	f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
	if err != nil {
		return -1
	}
	chunk := []byte("0123456789abcdef")
	for i := int64(0); i < chunks; i++ {
		f.Write(chunk)
	}
	f.Close()
	r, err := os.Open(path)
	if err != nil {
		return -2
	}
	var count int64 = 0
	b := make([]byte, 1)
	for {
		n, _ := r.Read(b)
		if n != 1 {
			break
		}
		count++
	}
	r.Close()
	os.Remove(path)
	return count
}

func main() {
	var result int64 = 0
	for r := int64(0); r < 200; r++ {
		result = roundTrip(15)
	}
	os.Exit(int(result & 0xff))
}

File: benchmarks/runtime/programs/c/file-io.c
/* The C peer of file-io.tuo: the identical open/write/read/close/unlink
 * sequence — 200 rounds of writing 15 sixteen-byte chunks and reading the
 * 240 bytes back one byte per read(2) call, matching the tuonelang program's
 * byte-at-a-time effect crossings. Exit byte: 240. */
#include <fcntl.h>
#include <unistd.h>

static long long round_trip(long long chunks) {
    const char *path = "file-io-bench.tmp";
    int fd = open(path, O_WRONLY | O_CREAT | O_TRUNC, 0644);
    if (fd < 0) {
        return -1;
    }
    for (long long i = 0; i < chunks; i++) {
        write(fd, "0123456789abcdef", 16);
    }
    close(fd);
    int rfd = open(path, O_RDONLY);
    if (rfd < 0) {
        return -2;
    }
    long long count = 0;
    unsigned char byte;
    while (read(rfd, &byte, 1) == 1) {
        count++;
    }
    close(rfd);
    unlink(path);
    return count;
}

int main(void) {
    long long result = 0;
    for (long long r = 0; r < 200; r++) {
        result = round_trip(15);
    }
    return (int)(result & 0xff);
}
